### ISSUE TITLE
Handling of arrays in SMTWriter

### ISSUE THREAD

**Issue by jpaykin** [GaloisInc/crucible#99]
`SMTWriter` is incorrectly converting array expressions into solver-specific representations of arrays.

- The `useSymbolicArray` flag is set incorrectly for yices;
- This flag is not consistently checked for e.g. in the `ConstantArray` case of `appSMTExpr`
- The library should have greater support for arrays being represented as functions in yices, perhaps as a flag for the existence of first-class higher-order functions

**Comment by robdockins** [comment]
A little more color here: the `SMTWriter` interface tries to take what4 expressions and render them into solver expressions in the "best possible" way.  The handling of arrays is one of those areas where the capabilities of solvers vary widely, so there are lots of alternate encodings and possible error conditions to account for.

In particular SMTLib2-compilant solvers have extremely limited ability to create and manipulate arrays.  Sometimes, we want to express operations that only can be expressed on functions; in this case we fall back on representing arrays as functions.  HOWEVER, SMTLib2 solvers do not support higher-order functions, so arrays encoded in this way cannot be passed as arguments to other functions.  There is some rather complicated logic that tries to figure out wether the desired expression can be encoded at all.

The native Yices language, on the other hand, has higher order functions, and no separate theory of arrays at all, just a distinguished combinator for doing point updates on functions.  Thus, array expressions sent to Yices should always be rendered as functions.  However, they don't have the same restrictions as SMTLib2 functions, and can appear as arguments to other functions, so the check should be relaxed in that case.

**Comment by joehendrix** [comment]
I would be careful with Yices -- it doesn't syntactically distinguish between arrays and functions, but you only can assert equality/inequality between Yices "functions" if they are constants or expressions constructed with the array update primitive.

I'm not sure if there are similar restrictions on passing functions as arguments; that may be more general, but I think if one does generalize this code, additional tests should be added to make sure you can use the features and still check satisfiability.

**Comment by jpaykin** [comment]
I've been working conservatively on this issue, and I just submitted a pull request (#105) that just addresses the first two points. Maybe in response to @joehendrix's comment, the third is not a priority right now?

**Comment by robdockins** [comment]
If you aren't blocked on this issue currently, I'm going to close it.  Feel free to reopen if this continues to cause problems.
